lib/mix/lib/mix/tasks/test.ex

```diff
@@ -14,17 +16,91 @@ defmodule Mix.Tasks.Test do
           Mix.raise("Failed to cover compile directory: " <> compile_path)
       end
 
-      output = opts[:output]
-
       fn ->
-        Mix.shell().info("\nGenerating cover results ...")
-        File.mkdir_p!(output)
+        Mix.shell().info("\nGenerating cover results ...\n")
+        {:result, ok, _fail} = :cover.analyse(:coverage, :module)
+
+        results = Enum.sort_by(ok, &percentage(elem(&1, 1)), &>=/2)
+
+        if summary_opts = Keyword.get(opts, :summary, true), do: console(results, summary_opts)
+
+        html(results, opts)
+      end
+    end
+
+    defp console(results, true), do: console(results, [])
 
-        Enum.each(:cover.modules(), fn mod ->
-          {:ok, _} = :cover.analyse_to_file(mod, '#{output}/#{mod}.html', [:html])
+    defp console(results, opts) when is_list(opts) do
+      Mix.shell().info("Percentage | Module")
+      Mix.shell().info("-----------|--------------------------")
+
+      total =
+        results
+        |> Stream.each(&display(&1, opts))
+        |> Enum.reduce({0, 0}, fn {_, {cov, not_cov}}, {tot_cov, tot_not_cov} ->
+          {tot_cov + cov, tot_not_cov + not_cov}
         end)
+
+      Mix.shell().info("-----------|--------------------------")
+
+      display({"Total", total}, opts)
+      Mix.shell().info("")
+    end
+
+    defp console(_, _), do: nil
+
+    defp html(results, opts) do
+      output = opts[:output]
+      File.mkdir_p!(output)
+
+      for {mod, _} <- results do
+        {:ok, _} = :cover.analyse_to_file(mod, '#{output}/#{mod}.html', [:html])
       end
+
+      Mix.shell().info([
+        "Generated HTML coverage results in '",
+        output,
+        "' directory\n"
+      ])
     end
+
+    defp colour(percentage, true), do: colour(percentage, @default_threshold)
+    defp colour(_, false), do: ""
+    defp colour(percentage, threshold) when percentage > threshold, do: :green
+    defp colour(_, _), do: :red
+
+    defp display({name, coverage}, opts) do
+      threshold = Keyword.get(opts, :threshold, @default_threshold)
+      percentage = percentage(coverage)
+
+      Mix.shell().info([
+        colour(percentage, threshold),
+        format(percentage, 9),
+        "%",
+        :reset,
+        " | ",
+        format_name(name)
+      ])
+    end
+
+    defp percentage({0, 0}), do: 100
+    defp percentage({cov, not_cov}), do: cov / (cov + not_cov) * 100
+
+    defp format(num, len) when is_integer(num) do
+      num
+      |> Integer.to_string()
+      |> String.pad_leading(len)
+    end
+
+    defp format(num, len) when is_float(num) do
+      num
+      |> Float.round(2)
+      |> Float.to_string()
+      |> String.pad_leading(len)
+    end
```

Comment:
Could we use `:io_lib.format("~#{len}.2f", [num])` for formatting here? This will always give us 2 fields of precision and nicely align the results.